### PR TITLE
Support for constant arrays of structured type

### PR DESCRIPTION
**Pull request by gmarkall**
This PR fixes issue #897 by adding support for constant arrays of complex values and structured arrays. (Updated from earlier, which only supported complex types, as the amount of work required was smaller than I anticipated).

### CODE REVIEW

**Comment by gmarkall** [comment]
On Python 2, the array data is:

```LLVM
@".const.array.data" = internal  constant [10 x [8 x i8]] [[8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 , i8 ?], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 �, i8 ?], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8  , i8 @], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 `, i8 @], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 �, i8 @], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 �, i8 @], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 �, i8 @], [8 x i8] [i8 , i8 , i8 , i8 , i8 , i8 , i8 �, i8 @], [8 x i8] [i8, i8 , i8 , i8 , i8 , i8 , i8, i8 A], [8 x i8] [i8         , i8 , i8 , i8 , i8 , i8 , i8 , i8 A]]
```

On Python 3, it is:

```LLVM
@".const.array.data" = internal  constant [10 x [8 x i8]] [[8 x i8] [i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 63], [8 x i8] [i8 1, i8 0, i8 0, i8 0, i8 0, i8 0, i8 192, i8 63], [8 x i8] [i8 2, i8 0, i8 0, i8 0, i8 0, i8 0, i8 32, i8 64], [8 x i8] [i8 3, i8 0, i8 0, i8 0, i8 0, i8 0, i8 96, i8 64], [8 x i8] [i8 4, i8 0, i8 0, i8 0, i8 0, i8 0, i8 144, i8 64], [8 x i8] [i8 5, i8 0, i8 0, i8 0, i8 0, i8 0, i8 176, i8 64], [8 x i8] [i8 6, i8 0, i8 0, i8 0, i8 0, i8 0, i8 208, i8 64], [8 x i8] [i8 7, i8 0, i8 0, i8 0, i8 0, i8 0, i8 240, i8 64], [8 x i8] [i8 8, i8 0, i8 0, i8 0, i8 0, i8 0, i8 8, i8 65], [8 x i8] [i8 9, i8 0, i8 0, i8 0, i8 0, i8 0, i8 24, i8 65]]
```

Does anyone have any pointers as to why the 0s are missing on Python 2?

File: numba/targets/base.py [hunk 1]
```diff
@@ -431,8 +431,12 @@ def get_constant_struct(self, builder, ty, val):
                       for i, v in enumerate(val)]
             return Constant.struct(consts)
 
+        elif isinstance(ty, types.Record):
+            consts = [self.get_constant(types.int8, b) for b in val.tobytes()]
```
Comment (pitrou) [inline numba/targets/base.py]: tobytes() gives you a str object under Python 2, and iterating gives you one-character strings.
Also, tobytes() appeared in Numpy 1.9, you should use tostring() instead (which does the same thing).
You can for example cast to bytearray:
```
>>> rec_X = np.recarray(10, dtype=np.dtype([('a', np.int32), ('b', np.float32)]))
>>> list(bytearray(rec_X.tostring()))
[224, 24, 169, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 176, 234, 167, 1, 0, 0, 0, 0, 144, 243, 94, 1, 0, 0, 0, 0, 112, 58, 176, 1, 0, 0, 0, 0, 208, 70, 180, 1, 0, 0, 0, 0, 46, 105, 110, 116, 51, 50, 41, 44, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 111, 97, 116, 51, 128, 53, 134, 90, 244, 127, 0, 0]
```

Comment (pitrou) [inline numba/targets/base.py]: Actually, you should be able to pass an array directly as the constant value:
```
    return Constant.array(Type.int(8), bytearray(val.tostring()))
```

Comment (gmarkall) [inline numba/targets/base.py]: Ah, many thanks for this!

**Comment by pitrou** [comment]
Looks good, thank you.

**Pull request merged**